File: src/test/java/easy/SingleNumberTest.java
package easy;

import static org.junit.jupiter.api.Assertions.*;

import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class SingleNumberTest {
  SingleNumber instance = new SingleNumber();

  @ParameterizedTest
  @MethodSource("dataSingleNumber")
  public void testSingleNumber(int[] nums, int expected) {

    int result = instance.solve(nums);

    assertEquals(expected, result);
  }

  public static Stream<Arguments> dataSingleNumber() {
    return Stream.of(
        Arguments.of(new int[] {2, 2, 1}, 1),
        Arguments.of(new int[] {4, 1, 2, 1, 2}, 4),
        Arguments.of(new int[] {1}, 1));
  }
}

File: easy/single_number_test.go
package easy

import (
	"testing"
)

func TestSingleNumber(t *testing.T) {
	cases := []struct {
		nums     []int
		expected int
	}{
		{[]int{2, 2, 1}, 1},
		{[]int{4, 1, 2, 1, 2}, 4},
		{[]int{1}, 1},
	}
	for _, c := range cases {
		t.Run("", func(t *testing.T) {
			result := SingleNumber(c.nums)
			if result != c.expected {
				t.Errorf("SingleNumber(%v) == %v, expected %v", c.nums, result, c.expected)
			}

		})

	}
}
